policy/policy-data.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package policy
+
+import (
+	"fmt"
+	"strings"
+
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/sa"
+
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+)
+
+const whitelisted = "whitelist"
+const blacklisted = "blacklist"
+
+// DomainRule ...
+type DomainRule struct {
+	Rule string `db:"rule"`
+	Type string `db:"type"`
+}
+
+func reverseName(domain string) string {
+	labels := strings.Split(domain, ".")
+	for i, j := 0, len(labels)-1; i < j; i, j = i+1, j-1 {
+		labels[i], labels[j] = labels[j], labels[i]
+	}
+	return strings.Join(labels, ".")
+}
+
+// PolicyAuthorityDatabaseImpl enforces policy decisions based on various rule
+// lists
+type PolicyAuthorityDatabaseImpl struct {
+	log   *blog.AuditLogger
+	dbMap *gorp.DbMap
+}
+
+// NewPolicyAuthorityDatabaseImpl constructs a Policy Authority Database (and
+// creates tables if they are non-existent)
+func NewPolicyAuthorityDatabaseImpl(driver, name string) (padb *PolicyAuthorityDatabaseImpl, err error) {
+	logger := blog.GetAuditLogger()
+	dbMap, err := sa.NewDbMap(driver, name)
+	if err != nil {
+		return nil, err
+	}
+
+	dbMap.AddTableWithName(DomainRule{}, "ruleList").SetKeys(false, "Rule")
+
+	err = dbMap.CreateTablesIfNotExists()
+	if err != nil {
+		return
+	}
+
+	padb = &PolicyAuthorityDatabaseImpl{
+		dbMap: dbMap,
+		log:   logger,
+	}
+
+	return padb, nil
+}
+
+// LoadRules loads the whitelist and blacklist into the database in a transaction
+// deleting any previous content
+func (padb *PolicyAuthorityDatabaseImpl) LoadRules(rules []DomainRule) error {
+	tx, err := padb.dbMap.Begin()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	_, err = tx.Exec("DELETE FROM ruleList")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	for _, r := range rules {
+		r.Rule = reverseName(r.Rule)
+		tx.Insert(&r)
+	}
+
+	err = tx.Commit()
+	return err
+}
+
+// DumpRules retrieves all DomainRules in the database so they can be written to
+// disk
+func (padb *PolicyAuthorityDatabaseImpl) DumpRules() ([]DomainRule, error) {
+	var dR []DomainRule
+	_, err := padb.dbMap.Select(&dR, "SELECT * FROM ruleList")
+	for _, r := range dR {
+		r.Rule = reverseName(r.Rule)
+	}
+
+	return dR, err
+}
+
+// CheckRules will query the database for white/blacklist rules that match host,
+// if both whitelist and blacklist rules are found the whitelist will always win
+func (padb *PolicyAuthorityDatabaseImpl) CheckRules(host string, requireWhitelisted bool) error {
+	host = reverseName(host)
+	var rules []DomainRule
+	_, err := padb.dbMap.Select(
+		&rules,
+		`SELECT type,rule FROM ruleList WHERE :host >= rule ORDER BY rule ASC`,
```

Comment:
Why is >= used here? This seems like it would capture every single rule where the string host is "larger than" which could be the entire dataset?

Also, if you want to indexing to be fast, reverse the order of the domain splits.

That is, "foo.example.com" becomes "com.example.foo". This means you can do rule LIKE "reversed_host%".